src/Compilers/CSharp/Portable/Symbols/Source/SourceOrdinaryMethodSymbolBase.cs

```diff
@@ -0,0 +1,668 @@
+                                report = ErrorCode.ERR_OverrideValConstraintNotSatisfied;
+                                break;
+                            }
+                            continue;
+                        default:
+                            continue;
+                    }
+
+                    diagnostics.Add(report, _typeParameters[i].Locations[0], this, _typeParameters[i],
+                                    overriddenOrExplicitlyImplementedMethod.TypeParameters[i], overriddenOrExplicitlyImplementedMethod);
+                }
+            }
+
+            CheckModifiers(MethodKind == MethodKind.ExplicitInterfaceImplementation, isVararg, HasAnyBody, location, diagnostics);
+
+            return;
+        }
+
+        protected abstract (TypeWithAnnotations ReturnType, ImmutableArray<ParameterSymbol> Parameters, bool IsVararg, ImmutableArray<TypeParameterConstraintClause> declaredConstraintsForOverrideOrImplement) MakeParametersAndBindReturnType(DiagnosticBag diagnostics);
+
+        protected abstract void ExtensionMethodChecks(DiagnosticBag diagnostics);
+
+        protected abstract MethodSymbol FindExplicitlyImplementedMethod(DiagnosticBag diagnostics);
+
+        protected abstract Location ReturnTypeLocation { get; }
+
+        protected abstract TypeSymbol ExplicitInterfaceType { get; }
+
+        protected abstract bool HasAnyBody { get; }
+
+        protected sealed override void LazyAsyncMethodChecks(CancellationToken cancellationToken)
+        {
+            Debug.Assert(this.IsPartial == state.HasComplete(CompletionPart.FinishMethodChecks),
+                "Partial methods complete method checks during construction.  " +
+                "Other methods can't complete method checks before executing this method.");
+
+            if (!this.IsAsync)
+            {
+                CompleteAsyncMethodChecks(diagnosticsOpt: null, cancellationToken: cancellationToken);
+                return;
+            }
+
+            DiagnosticBag diagnostics = DiagnosticBag.GetInstance();
+            AsyncMethodChecks(diagnostics);
+
+            CompleteAsyncMethodChecks(diagnostics, cancellationToken);
+            diagnostics.Free();
+        }
+
+        private void CompleteAsyncMethodChecks(DiagnosticBag diagnosticsOpt, CancellationToken cancellationToken)
+        {
+            if (state.NotePartComplete(CompletionPart.StartAsyncMethodChecks))
+            {
+                if (diagnosticsOpt != null)
+                {
+                    AddDeclarationDiagnostics(diagnosticsOpt);
+                }
+
+                CompleteAsyncMethodChecksBetweenStartAndFinish();
+                state.NotePartComplete(CompletionPart.FinishAsyncMethodChecks);
+            }
+            else
+            {
+                state.SpinWaitComplete(CompletionPart.FinishAsyncMethodChecks, cancellationToken);
+            }
+        }
+
+        protected virtual void CompleteAsyncMethodChecksBetweenStartAndFinish()
```

Comment:
nit: Consider `abstract` and when the new derived type comes, it can have an empty body (won't need to raise event for declared symbol).